geo-rep: ensure py2py3 compatibily

Shwetha-Acharya opened the pull request:
Problem: decode() returns unicode value on python2 and class bytes value on python3.

               In python3, split returns:- TypeError: a bytes-like object is required, not 'str'
               when a str seperator '\n' is used, as python3 expects seperator also to be bytes
               literal line to be split is bytes literal

               Traceback observed is as follows:
         
               Traceback (most recent call last):
                   File "/usr/libexec/glusterfs/python/syncdaemon/gsyncd.py", line 317, in main
                   func(args)
                   File "/usr/libexec/glusterfs/python/syncdaemon/subcmds.py", line 60, in subcmd_monitor
                   return monitor.monitor(local, remote)
                   File "/usr/libexec/glusterfs/python/syncdaemon/monitor.py", line 360, in monitor
                   return Monitor().multiplex(*distribute(local, remote))
                   File "/usr/libexec/glusterfs/python/syncdaemon/monitor.py", line 319, in distribute
                   svol = Volinfo(slave.volume, "localhost", prelude, master=False)
                   File "/usr/libexec/glusterfs/python/syncdaemon/syncdutils.py", line 924, in __init__
                   po.terminate_geterr()
                   File "/usr/libexec/glusterfs/python/syncdaemon/syncdutils.py", line 894, in terminate_geterr
                   self.errfail()
                   File "/usr/libexec/glusterfs/python/syncdaemon/syncdutils.py", line 863, in errfail
                   self.errlog()
                   File "/usr/libexec/glusterfs/python/syncdaemon/syncdutils.py", line 853, in errlog
                   ls = l.split('\n')
                   TypeError: a bytes-like object is required, not 'str'

Solution: when the str seperator is converted to bytes literal, both python2 and python3
               codebase work seemlessly.

               Python2 also supports split, with bytes literal as a seperator and line to be
               split is unicode.

Fixes: #2469
Change-Id: I0b383ef320e75fb697870b105ad72265b837234f
Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya committed:
geo-rep: ensure py2py3 compatibily

Problem: decode() returns unicode value on python2 and class bytes value on python3.

         In python3, split returns:- TypeError: a bytes-like object is required, not 'str'
         when a str seperator '\n' is used, as python3 expects seperator also to be bytes
         literal line to be split is bytes literal

         Traceback observed is as follows:

         Traceback (most recent call last):
           File "/usr/libexec/glusterfs/python/syncdaemon/gsyncd.py", line 317, in main
           func(args)
           File "/usr/libexec/glusterfs/python/syncdaemon/subcmds.py", line 60, in subcmd_monitor
           return monitor.monitor(local, remote)
           File "/usr/libexec/glusterfs/python/syncdaemon/monitor.py", line 360, in monitor
           return Monitor().multiplex(*distribute(local, remote))
           File "/usr/libexec/glusterfs/python/syncdaemon/monitor.py", line 319, in distribute
           svol = Volinfo(slave.volume, "localhost", prelude, master=False)
           File "/usr/libexec/glusterfs/python/syncdaemon/syncdutils.py", line 924, in __init__
           po.terminate_geterr()
           File "/usr/libexec/glusterfs/python/syncdaemon/syncdutils.py", line 894, in terminate_geterr
           self.errfail()
           File "/usr/libexec/glusterfs/python/syncdaemon/syncdutils.py", line 863, in errfail
           self.errlog()
           File "/usr/libexec/glusterfs/python/syncdaemon/syncdutils.py", line 853, in errlog
           ls = l.split('\n')
           TypeError: a bytes-like object is required, not 'str'

Solution: when the str seperator is converted to bytes literal, both python2 and python3
          codebase work seemlessly.

          Python2 also supports split, with bytes literal as a seperator and line to be
          split is unicode.

Fixes: #2469
Change-Id: I0b383ef320e75fb697870b105ad72265b837234f
Signed-off-by: Shwetha K Acharya <sacharya@redhat.com>

Shwetha-Acharya commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/bugs/glusterd/brick-order-check-add-brick.t

0 test(s) generated core 


6 test(s) needed retry 
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
./tests/basic/afr/gfid-mismatch-resolution-with-cli.t
./tests/bugs/changelog/bug-1208470.t
./tests/bugs/cli/bug-1169302.t
./tests/bugs/glusterd/brick-order-check-add-brick.t

2 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
https://build.gluster.org/job/gh_centos7-regression/2023/

Shwetha-Acharya commented:
/run regression

aravindavk commented on geo-replication/syncdaemon/syncdutils.py:
```diff
@@ -850,8 +850,13 @@ def errlog(self):
         for l in self.elines:
-            ls = l.split('\n')
-            ls[0] = lp + ls[0]
+            #handle py2py3 compatibility
```
Try to fix where lines are added to `elines`(Ex: Line 888 `elines.append(b.decode())` (Line 893 with new changes))

Python3 expects bytes literal if the line is bytes literal. For example: Both `b"A:B".split(b":")` and `"A:B".split(":")` are valid in Python3

Also bytes literal is added to Python 2.6 onwards. It does nothing but the syntax is supported. `b"A:B".split(b":")` works in >Python 2.6

aravindavk commented:
Traceback in the commit message looks different than the one it is fixing.

Shwetha-Acharya commented:
> Traceback in the commit message looks different than the one it is fixing.

sorry, wrong traceback was added in commit message

Shwetha-Acharya commented:
@aravindavk 
[root@sacharya ~]# python2
Python 2.7.18 (default, Jan 22 2021, 11:25:37) 
[GCC 8.4.1 20200928 (Red Hat 8.4.1-1)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> path = "/home/sacharya/Downloads/glusterd-slave.log"
>>> fd = os.open(path, os.O_RDONLY)
>>> n = 1024
>>> readBytes = os.read(fd, n)
>>> readBytes.decode()
u''
>>> print(type(readBytes))
<type 'unicode'>
>>> ls = readBytes.split(b'\n')
>>> lp = ls.pop()


python3:

[root@sacharya ~]# python3
Python 3.6.8 (default, Mar 18 2021, 08:58:41) 
[GCC 8.4.1 20200928 (Red Hat 8.4.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> path = "/home/sacharya/Downloads/glusterd-slave.log"
>>> fd = os.open(path, os.O_RDONLY)
>>> n = 1024
>>> readBytes = os.read(fd, n)
>>> readBytes.decode()
''
>>> print(type(readBytes))
<class 'bytes'>
>>> ls = readBytes.split(b'\n')
>>> lp = ls.pop()
>>>

Shwetha-Acharya commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/bugs/glusterd/brick-order-check-add-brick.t

0 test(s) generated core 


8 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/basic/ec/ec-1468261.t
./tests/basic/ec/ec-data-heal.t
./tests/basic/glusterd/heald.t
./tests/bugs/glusterd/brick-order-check-add-brick.t

2 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
https://build.gluster.org/job/gh_centos7-regression/2031/

Shwetha-Acharya commented:
/run regression
